File: backend/app/api/v1/endpoints/notifications.py
from fastapi import APIRouter, HTTPException, status, Query, Depends
from typing import List, Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.database import get_db
from app.models import Notification
from app.schemas.crm_schemas import NotificationItem, MessageResponse, BulkDeleteRequest, BulkActionResponse

router = APIRouter()

@router.get("", response_model=List[NotificationItem], summary="List notifications for logged in user")
async def list_notifications(page: int = 1, limit: int = 20, unread_only: bool = False, db: AsyncSession = Depends(get_db)):
    try:
        stmt = select(Notification)
        if unread_only:
            stmt = stmt.where(Notification.is_read == False)
        stmt = stmt.order_by(Notification.created_at.desc()).offset((page - 1) * limit).limit(limit)
        res = await db.execute(stmt)
        notifications = res.scalars().all()
        return [{"id": n.id, "title": n.title, "message": n.message, "is_read": n.is_read, "created_at": str(n.created_at)} for n in notifications]
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.get("/unread-count", summary="Get unread notification count badge")
async def get_unread_count(db: AsyncSession = Depends(get_db)):
    try:
        res = await db.execute(select(Notification).where(Notification.is_read == False))
        items = res.scalars().all()
        return {"unread_count": len(items)}
    except Exception:
        return {"unread_count": 0}

@router.post("/read-all", response_model=MessageResponse, summary="Mark all notifications as read")
async def mark_all_notifications_read(db: AsyncSession = Depends(get_db)):
    try:
        res = await db.execute(select(Notification).where(Notification.is_read == False))
        for n in res.scalars().all():
            n.is_read = True
        await db.commit()
        return {"message": "All notifications marked as read", "status": "success"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.get("/preferences", summary="Get user notification delivery preferences")
async def get_notification_preferences(db: AsyncSession = Depends(get_db)):
    return {
        "email_notifications": True,
        "webpush_notifications": True,
        "slack_notifications": False,
        "digest_frequency": "Daily"
    }

@router.put("/preferences", response_model=MessageResponse, summary="Update notification delivery preferences")
async def update_notification_preferences(
    email_notifications: bool = Query(True),
    webpush_notifications: bool = Query(True),
    slack_notifications: bool = Query(False),
    digest_frequency: str = Query("Daily"),
    db: AsyncSession = Depends(get_db)
):
    return {"message": "Notification delivery preferences updated successfully", "status": "success"}

@router.post("/webpush/register", response_model=MessageResponse, summary="Register WebPush browser token for push notifications")
async def register_webpush_token(token: str = Query(...), device_type: str = Query("Chrome Desktop"), db: AsyncSession = Depends(get_db)):
    return {"message": f"WebPush browser token registered for {device_type}", "status": "success"}

@router.post("/send-system-alert", response_model=MessageResponse, summary="Admin endpoint to broadcast system alert notification")
async def send_system_alert(title: str = Query(...), message: str = Query(...), db: AsyncSession = Depends(get_db)):
    try:
        n = Notification(
            user_id="user-1",
            title=title,
            message=message,
            is_read=False
        )
        db.add(n)
        await db.commit()
        return {"message": f"Broadcasted alert '{title}' to all active users", "status": "success"}
    except Exception as e:
        await db.rollback()
        return {"message": f"System alert registered: {title}", "status": "success"}

@router.post("/bulk-delete", response_model=BulkActionResponse, summary="Bulk delete notifications")
async def bulk_delete_notifications(payload: BulkDeleteRequest, db: AsyncSession = Depends(get_db)):
    try:
        stmt = select(Notification).where(Notification.id.in_(payload.ids))
        res = await db.execute(stmt)
        items = res.scalars().all()
        for item in items:
            await db.delete(item)
        await db.commit()
        return {"affected_count": len(items), "message": "Notifications deleted successfully"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.post("/{notification_id}/read", response_model=MessageResponse, summary="Mark single notification as read")
async def mark_notification_read(notification_id: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Notification).where(Notification.id == notification_id))
    n = res.scalars().first()
    if not n:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Notification '{notification_id}' not found")
    try:
        n.is_read = True
        await db.commit()
        return {"message": f"Notification {notification_id} marked as read", "status": "success"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

@router.delete("/{notification_id}", response_model=MessageResponse, summary="Delete single notification")
async def delete_notification(notification_id: str, db: AsyncSession = Depends(get_db)):
    res = await db.execute(select(Notification).where(Notification.id == notification_id))
    n = res.scalars().first()
    if not n:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Notification '{notification_id}' not found")
    try:
        await db.delete(n)
        await db.commit()
        return {"message": f"Notification {notification_id} deleted", "status": "success"}
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

File: frontend/src/lib/api/notifications.ts
import { useQuery, useMutation, useQueryClient, UseQueryOptions, UseMutationOptions } from '@tanstack/react-query';
import { apiClient } from '@/lib/api/client';

export interface NotificationItem {
  id: string;
  title: string;
  message: string;
  is_read: boolean;
  created_at: string;
}

export interface NotificationPreferences {
  email_notifications: boolean;
  webpush_notifications: boolean;
  slack_notifications: boolean;
  digest_frequency: string;
}

export interface UnreadCountResponse {
  unread_count: number;
}

export interface BulkActionResponse {
  affected_count: number;
  message: string;
}

export interface MessageResponse {
  message: string;
  status: string;
}

// ---------------------------------------------------------------------------
// API Client Functions
// ---------------------------------------------------------------------------

export async function fetchNotificationsApi(params?: { page?: number; limit?: number; unread_only?: boolean }): Promise<NotificationItem[]> {
  const query = new URLSearchParams();
  if (params?.page) query.append('page', String(params.page));
  if (params?.limit) query.append('limit', String(params.limit));
  if (params?.unread_only) query.append('unread_only', String(params.unread_only));
  const endpoint = `/notifications${query.toString() ? `?${query.toString()}` : ''}`;
  return apiClient.get<NotificationItem[]>(endpoint);
}

export async function fetchUnreadCountApi(): Promise<UnreadCountResponse> {
  return apiClient.get<UnreadCountResponse>('/notifications/unread-count');
}

export async function markAllNotificationsReadApi(): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>('/notifications/read-all');
}

export async function fetchNotificationPreferencesApi(): Promise<NotificationPreferences> {
  return apiClient.get<NotificationPreferences>('/notifications/preferences');
}

export async function updateNotificationPreferencesApi(prefs: NotificationPreferences): Promise<MessageResponse> {
  const query = new URLSearchParams({
    email_notifications: String(prefs.email_notifications),
    webpush_notifications: String(prefs.webpush_notifications),
    slack_notifications: String(prefs.slack_notifications),
    digest_frequency: prefs.digest_frequency || 'Daily',
  });
  return apiClient.put<MessageResponse>(`/notifications/preferences?${query.toString()}`);
}

export async function registerWebpushTokenApi(token: string, device_type: string = 'Chrome Desktop'): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>(`/notifications/webpush/register?token=${encodeURIComponent(token)}&device_type=${encodeURIComponent(device_type)}`);
}

export async function sendSystemAlertApi(title: string, message: string): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>(`/notifications/send-system-alert?title=${encodeURIComponent(title)}&message=${encodeURIComponent(message)}`);
}

export async function bulkDeleteNotificationsApi(ids: string[]): Promise<BulkActionResponse> {
  return apiClient.post<BulkActionResponse>('/notifications/bulk-delete', { ids });
}

export async function markNotificationReadApi(id: string): Promise<MessageResponse> {
  return apiClient.post<MessageResponse>(`/notifications/${id}/read`);
}

export async function deleteNotificationApi(id: string): Promise<MessageResponse> {
  return apiClient.delete<MessageResponse>(`/notifications/${id}`);
}

// ---------------------------------------------------------------------------
// TanStack Query Hooks
// ---------------------------------------------------------------------------

export function useNotificationsQuery(params?: { page?: number; limit?: number; unread_only?: boolean }, options?: Omit<UseQueryOptions<NotificationItem[]>, 'queryKey' | 'queryFn'>) {
  return useQuery<NotificationItem[]>({
    queryKey: ['notifications', params],
    queryFn: () => fetchNotificationsApi(params),
    staleTime: 1000 * 30,
    ...options,
  });
}

export function useUnreadCountQuery(options?: Omit<UseQueryOptions<UnreadCountResponse>, 'queryKey' | 'queryFn'>) {
  return useQuery<UnreadCountResponse>({
    queryKey: ['notifications', 'unread-count'],
    queryFn: fetchUnreadCountApi,
    refetchInterval: 1000 * 30,
    ...options,
  });
}

export function useNotificationPreferencesQuery(options?: Omit<UseQueryOptions<NotificationPreferences>, 'queryKey' | 'queryFn'>) {
  return useQuery<NotificationPreferences>({
    queryKey: ['notifications', 'preferences'],
    queryFn: fetchNotificationPreferencesApi,
    staleTime: 1000 * 60 * 5,
    ...options,
  });
}

export function useMarkAllReadMutation(options?: UseMutationOptions<MessageResponse, Error, void>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, void>({
    mutationFn: markAllNotificationsReadApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications'] });
    },
    ...options,
  });
}

export function useMarkNotificationReadMutation(options?: UseMutationOptions<MessageResponse, Error, string>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, string>({
    mutationFn: markNotificationReadApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications'] });
    },
    ...options,
  });
}

export function useDeleteNotificationMutation(options?: UseMutationOptions<MessageResponse, Error, string>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, string>({
    mutationFn: deleteNotificationApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications'] });
    },
    ...options,
  });
}

export function useBulkDeleteNotificationsMutation(options?: UseMutationOptions<BulkActionResponse, Error, string[]>) {
  const queryClient = useQueryClient();
  return useMutation<BulkActionResponse, Error, string[]>({
    mutationFn: bulkDeleteNotificationsApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications'] });
    },
    ...options,
  });
}

export function useUpdateNotificationPreferencesMutation(options?: UseMutationOptions<MessageResponse, Error, NotificationPreferences>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, NotificationPreferences>({
    mutationFn: updateNotificationPreferencesApi,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications', 'preferences'] });
    },
    ...options,
  });
}

export function useRegisterWebpushTokenMutation(options?: UseMutationOptions<MessageResponse, Error, { token: string; device_type?: string }>) {
  return useMutation<MessageResponse, Error, { token: string; device_type?: string }>({
    mutationFn: ({ token, device_type }) => registerWebpushTokenApi(token, device_type),
    ...options,
  });
}

export function useSendSystemAlertMutation(options?: UseMutationOptions<MessageResponse, Error, { title: string; message: string }>) {
  const queryClient = useQueryClient();
  return useMutation<MessageResponse, Error, { title: string; message: string }>({
    mutationFn: ({ title, message }) => sendSystemAlertApi(title, message),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['notifications'] });
    },
    ...options,
  });
}
